lib/polo/collector.rb

```diff
@@ -14,15 +14,25 @@ def initialize(base_class, id, dependency_tree={})
     # ActiveSupport::Notifications block and collecting every generate SQL query.
     #
     def collect
-      unprepared_statement do
-        ActiveSupport::Notifications.subscribed(collector, 'sql.active_record') do
-          base_finder = @base_class.includes(@dependency_tree).where(@base_class.primary_key => @id)
-          collect_sql(@base_class, base_finder.to_sql)
-          base_finder.to_a
+      base_finder = @base_class.includes(@dependency_tree).where(@base_class.primary_key => @id)
+      enumerable = Enumerator.new do |yielder|
+        collect_sql(@base_class, base_finder.to_sql)
+        unprepared_statement do
+          ActiveSupport::Notifications.subscribed(collector, 'sql.active_record') do
+            base_finder.find_in_batches.with_index do |batch, batch_index|
+              # Expose each select to the enumerator
+              yielder.yield(@selects)
+              # Now reset the accumulator (don't hog memory!)
+              @selects = Set.new
+            end
+          end
         end
       end
-
-      @selects.compact.uniq
```

Comment:
Another potential breaking change, we're no longer uniq-ing the whole set. This is the main performance offender, since you can't  I don't make unique an unordered set without having the whole set in memory.  However you already have an API to handle conflicts, so I don't _think_ this is an issue? For us the bigger issue is the memory tradeoff, anyway.